### ISSUE TITLE
[Proposal]: Extension everything!

### ISSUE THREAD

**Issue by VBAndCs** [dotnet/csharplang#3921]
# Extension everything

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
There was many requests for extension properties and operators. I think it can be done by allowing extension partial classes, like this:

## Detailed design
[design]: #detailed-design

```C#
class Foo
{
        public int Value { get; set; }
}

[Extension]
partial class Foo
{
  public void SetValue(int value) => this.Value = value;

  public int Prop1 { set; get; }
  
   public static int operator +(Foo foo1, Foo foo2) => foo1.Value + foo2.Value;  
            
}
```
Note that C# should allow using the partial keyword with the `[Extension]` attr, dispite the original Foo class doesn’t use the partial keyword, as we can extend a framework class. 

The extension partial can be lowered to:
```C#
class Foo_Extension
{
        Foo _foo;
        public Foo_Extension(Foo foo) => _foo = foo;

        public void SetValue(int value) => _foo.Value = value;

        public int Prop1 { set; get; }

        public static int operator +(Foo_Extension foo_ext1, Foo_Extension foo_ext2) => foo_ext1._foo.Value + foo_ext2._foo.Value;  
            
}
```

Of course, this way can't work with value types, unless c# allows to set them by  ref in the constructor:
```
  public Foo_Extension(ref Foo foo) => ref _foo = ref foo;
```

The compiler should in create an instance of the Foo_Extension class whenever one of its members are called through the Foo class:

```C#
var f1 = new Foo();
f1.SetValue(5);
var f2 = new Foo();
f2.SetValue(10);
f2.Prop1 = 1;

var y = f1 + f2;
```

Which can be lowered to:
```C#
var f1 = new Foo();
var f1_Ext = new Foo_Extension(f1);
f1_Ext.SetValue(5);
var f2 = new Foo();
var f2_Ext = new Foo_Extension(f2);
f2_Ext.SetValue(10);
f2_Ext.Prop1 = 1;

var y = f1_Ext + f2_Ext;
```

**Comment by HaloFour** [comment]
Your suggested approach would incur allocations for every extension at every call site.  There are existing proposals which are zero-cost abstractions and the team is actively looking into this as a part of the shapes/roles design.

**Comment by YairHalberstadt** [comment]
Note that this should really be opened as a discussion, as issues are only for full fledged formal proposals.

**Comment by VBAndCs** [comment]
@HaloFour
Not for every object. Only for those using the extensions. 
Attaching extension properties needs some instance. If there is no properties in the extension class, it can be reduces to normal extension methods.
In fact I have another idea for auto properties here: https://github.com/dotnet/vblang/issues/570.
But I faced a challenge of some object references (like strings) as I mentioned in gitter.

**Comment by VBAndCs** [comment]
@YairHalberstadt 
Your template disguided  me. I just followed the instructions.

**Comment by HaloFour** [comment]
@VBAndCs 

> Only for those using the extensions.

For every single instance, for every single extension you use, for every single time you use those extensions.  It's _way_ too easy for this to spin into a _lot_ of unnecessary allocations.  Using an extension in one method and then passing that instance to another method will result in multiple allocations for the same allocation.  Using an extension in a loop will result in allocations for every single iteration.

**Comment by HaloFour** [comment]
Here's the currently championed proposal that is looking into this space:  https://github.com/dotnet/csharplang/issues/192

**Comment by 333fred** [comment]
Closing as a duplicate. We already have #192 for discussion on improving extensions (and indeed we usually refer to that proposal as "Extension Everything" as well).

**Comment by VBAndCs** [comment]
> passing that instance to another method will result in multiple allocations for the same allocation.

Why? The ext holds a ref to the obj that is passed to the function, so, it is still valid across functions. No need to create a new ext obj.

> Using an extension in a loop will result in allocations for every single iteration.

Why? Once you created the ext, it will be alive as long as the obj, and will not be disposed unless the obj is disposed.
I see my design is easy and practical, and easy to use. May be I use the extends keyword instead of partial, but I am avoiding to use `this`, so we may need to use the `[Extension(typeof(some class))]` to link the ext to the original class in .Net Framework.
The other proposal have many limitation I don't have here.
In short: I am using something very close to inheritance (except we can't access private and protected members here).
So, I think you should study of my proposal carefully.

**Comment by YairHalberstadt** [comment]
> Your template disguided me. I just followed the instructions.

A full proposal must be much more detailed and specific. For example in your case it should state:

1. What is allowed in an extension class?
Can it implement interfaces, inherit classes, define constructors, define static members, define implicit/explicit conversions etc.
2. What precisely are the laws for lowering. You just mentioned:
> Once you created the ext, it will be alive as long as the obj, and will not be disposed unless the obj is disposed.
We need a precise spec for how this works.
3. How do you bring an extension into scope?
4. What happens if their are multiple extensions in scope?
5. How does overload resolution work with extensions?
6. Can the extension live in a different namespace to the original class? If so how do we know to which class it's extending?

And probably more things I can't think of.

I wouldn't expect you to have answers to all these questions yet - it would probably be a waste of time since you first want to check if the LDM is interested. However, till all these details are worked out it's best to open things as discussions. Once the proposal is more concrete it can be moved to an issue.

**Comment by HaloFour** [comment]
@VBAndCs 

> Why? The ext holds a ref to the obj that is passed to the function, so, it is still valid across functions. No need to create a new ext obj.

Because `FooExt` isn't `Foo` and you're not passing `FooExt` instances alongside each `Foo` instance when you call another method.  So if that other method needs to use an extension it too needs to create new extension instances.

> Why? Once you created the ext, it will be alive as long as the obj,

Because you have to create a `FooExt` for every single instance of the `obj`.  So if you're in a loop over `Foo` instances that a new `FooExt` per iteration.
